common/determined_common/schemas/extensions.py

```diff
@@ -0,0 +1,159 @@
+import os
+import jsonschema
+
+from determined_common.schemas import expconf
+
+
+def _evaluate_unionKey(key, instance):
+    if key is None:
+        return False
+    if isinstance(key, list):
+        return all(_evaluate_unionKey(k, instance) for k in key)
+    if isinstance(key, str):
+        if key == "always":
+            return True
+        if key == "never":
+            return False
+        if key.startswith("not:"):
+            return not _evaluate_unionKey(key[len("not:") :], instance)
+        if key.startswith("const:"):
+            eq_idx = key.index("=")
+            name = key[len("const:") : eq_idx]
+            value = key[eq_idx + 1 :]
+            if not isinstance(instance, dict):
+                return False
+            return instance.get(name) == value
+        if key.startswith("singleproperty:"):
+            name = key[len("singleproperty:") :]
+            if not isinstance(instance, dict):
+                return False
+            if len(instance) != 1:
+                return False
+            return name in instance
+        if key.startswith("type:"):
+            typ = key[len("type:") :]
+            assert typ in ["array", "object"]
+            if typ == "array":
+                return isinstance(instance, list)
+            if typ == "object":
+                return isinstance(instance, dict)
+        if key.startswith("hasattr:"):
+            attr = key[len("hasattr:") :]
+            return isinstance(attr, dict) and attr in instance
+    raise ValueError(f"invalid unionKey: {key}")
+
+
+def union(validator, det_one_of, instance, schema):
+    # like oneOf but with custom error handling
+    message = det_one_of.get("defaultMessage", "union failed to validate")
+
+    selected_errors = None
+    valid = []
+
+    for idx, item in enumerate(det_one_of["items"]):
+        errors = [e for e in validator.descend(instance, schema=item, schema_path=idx)]
+        if errors:
+            key = item["unionKey"]
+            if not selected_errors and _evaluate_unionKey(key, instance):
+                selected_errors = errors
+        else:
+            valid.append(item)
+
+    if len(valid) == 1:
+        # No errors.
+        return
+
+    if len(valid) > 1:
+        yield jsonschema.ValidationError(
+            f"bug in validation! Multiple schemas matched: {valid}"
+        )
+        return
+
+    if selected_errors:
+        yield from selected_errors
```

Comment:
im confused how we dont we not end up returning errors from the first thing in the one of every time.